controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -86,25 +85,49 @@ public String getScopeName() {
      * @return : future of whether it was done or not
      */
     @Override
-    public CompletableFuture<Boolean> create(final StreamConfiguration configuration, long createTimestamp) {
-        final Create create = new Create(createTimestamp, configuration);
+    public CompletableFuture<CreateStreamResponse> create(final StreamConfiguration configuration, long createTimestamp) {
 
         return checkScopeExists()
-                .thenCompose(x -> checkStreamExists(create))
-                .thenCompose(x -> storeCreationTime(create))
-                .thenCompose(x -> createConfiguration(create))
-                .thenCompose(x -> createState(State.CREATING))
-                .thenCompose(x -> createSegmentTable(create))
-                .thenCompose(x -> createNewEpoch(0))
-                .thenCompose(x -> {
-                    final int numSegments = create.getConfiguration().getScalingPolicy().getMinNumSegments();
-                    final byte[] historyTable = TableHelper.createHistoryTable(create.getCreationTime(),
-                            IntStream.range(0, numSegments).boxed().collect(Collectors.toList()));
+                .thenCompose(x -> checkStreamExists(configuration, createTimestamp))
+                .thenCompose(createStreamResponse -> storeCreationTimeIfAbsent(createStreamResponse.getTimestamp())
+                        .thenCompose(x -> createConfigurationIfAbsent(createStreamResponse.getConfiguration()))
+                        .thenCompose(x -> createStateIfAbsent(State.CREATING))
+                        .thenCompose(x -> createNewSegmentTable(createStreamResponse.getConfiguration(), createStreamResponse.getTimestamp()))
+                        .thenCompose(x -> getState().thenCompose(state -> {
+                            if (state.equals(State.CREATING)) {
+                                return createNewEpoch(0);
+                            } else {
+                                return CompletableFuture.completedFuture(null);
+                            }
+                        }))
+                        .thenCompose(x -> {
```

Comment:
All of the other elements in this chain have nice named methods so the sequence is easy to follow. Perhaps this block and the one above could be turned into private methods.